### PR TITLE
lib: lte_link_control: multiple registered LTE event handlers and better testability

### PR DESCRIPTION
**Pull request by jhirsi**
This PR adds for lib/lte_link_control:
- a support for multiple registered LTE event handlers
- changes to improve testability.

### CODE REVIEW

**Comment by jtguggedal** [comment]
Thanks for doing these changes, they're definitely needed.
Will the unit tests mentioned in this PR be added to NCS?

**Comment by trantanen** [comment]
> Will the unit tests mentioned in this PR be added to NCS?

Yes. I'm planning to mock at_cmd library in location library unit tests. There are first initial tests available in the draft PR: https://github.com/nrfconnect/sdk-nrf/pull/5341
Those tests are likely broken at the moment but I just demoed them a month ago or so to see how we could get them done nicely. I'll update and extend the tests once our implementation gets more stable.

**Review by trantanen — APPROVED** [review]

**Review by rlubos — COMMENTED** [review]
This needs a changelog entry, we modify the API here.

File: lib/lte_link_control/lte_lc_helpers.c [hunk 1]
```diff
@@ -22,6 +22,112 @@
...
+ * @brief Find the handler from the event handler list.
+ *
+ * @return The node or NULL if not found and its previous node in @p prev_out.
+ */
+static struct event_handler *event_handler_list_find_node(struct event_handler **prev_out,
+	lte_lc_evt_handler_t handler)
+{
+	struct event_handler *prev = NULL, *curr, *tmp;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&handler_list, curr, tmp, node) {
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply lib/lte_link_control/lte_lc_helpers.c]: OK, thanks. Started to use SYS_SLIST_FOR_EACH_CONTAINER in this case.

**Comment by jhirsi** [comment]
> This needs a changelog entry, we modify the API here.

Added an entry.

**Review by rlubos — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -61,6 +61,7 @@ nRF9160
   * :ref:`lte_lc_readme` library:
 
     * Changed the value of an invalid E-UTRAN cell ID from zero to UINT32_MAX for the LTE_LC_EVT_NEIGHBOR_CELL_MEAS event.
+    * Added a support for multiple LTE event handlers. Thus, deregistation is not anymore possible by using lte_lc_register_handler(NULL), and now on, there is own function in API for doing that: lte_lc_deregister_handler().
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:64]:
```suggestion
    * Added support for multiple LTE event handlers. Thus, deregistration is not possible by using lte_lc_register_handler(NULL) anymore and it is done by the :c:func:`lte_lc_deregister_handler` function in the API.
```

**Review by jtguggedal — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst (hunk 2 above)
Comment (jhirsi) [reply doc/nrf/releases/release-notes-changelog.rst]: Thanks @umapraseeda. That change is now done.

**Review by umapraseeda — APPROVED** [review]

File: lib/lte_link_control/lte_lc.c [hunk 3]
```diff
@@ -745,24 +745,28 @@ int lte_lc_init(void)
 void lte_lc_register_handler(lte_lc_evt_handler_t handler)
 {
 	if (handler == NULL) {
-		evt_handler = NULL;
-
-		LOG_INF("Previously registered handler (%p) deregistered",
-			handler);
-
+		LOG_INF("NULL as a handler received: Nothing to be done.\n"
+			"Deregister of the handler can done by using lte_lc_deregister_handler()");
```
Comment (tokangas) [inline lib/lte_link_control/lte_lc.c:749]:
```suggestion
			"The handler can be deregistered using lte_lc_deregister_handler()");
```

**Review by tokangas — APPROVED** [review]
Looks good!

File: lib/lte_link_control/lte_lc.c (hunk 3 above)
Comment (jhirsi) [reply lib/lte_link_control/lte_lc.c]: Changed.

**Comment by trantanen** [comment]
> > Will the unit tests mentioned in this PR be added to NCS?
> 
> Yes. I'm planning to mock at_cmd library in location library unit tests. There are first initial tests available in the draft PR: #5341 Those tests are likely broken at the moment but I just demoed them a month ago or so to see how we could get them done nicely. I'll update and extend the tests once our implementation gets more stable.

Due to at_monitor and nrf_modem_at API use in the future, these unit test changes will need to change. lte_lc work for taking these into use is done in https://github.com/nrfconnect/sdk-nrf/pull/5741. Maybe we could drop the unit test support commit from this PR as I'll need to figure out what changes are required to get at_monitor and nrf_modem_at mocking done.

**Pull request merged**